### ISSUE TITLE
Validate implementation by training on another dataset

### ISSUE THREAD

**Issue by vitalwarley** [vitalwarley/research#3]
- [x] CIFAR-10
  - [x] [SimpleModel](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html)
  - [x] My Model

**Comment by vitalwarley** [comment]
`SimpleModel` no CIFAR10.

![2022-04-03-161533_400x688_scrot](https://user-images.githubusercontent.com/6365065/161444266-39a3ca8f-f473-4b23-86df-ed88bd77de11.png)

No tutorial que linkei houve 54% de acurácia no test set, logo minha implementação está com bug.

**Comment by vitalwarley** [comment]
Comparar com implementação original

- [ ] [SimpleModel](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html#define-a-convolutional-neural-network) -- ver camadas e `forward` 
- [ ] Função de perda -- `CrossEntropyLoss`
- [ ] Otimizador -- `SGD`

**Comment by vitalwarley** [comment]
> * [ ]  [SimpleModel](https://pytorch.org/tutorials/beginner/blitz/cifar10_tutorial.html#define-a-convolutional-neural-network) -- ver camadas e `forward`

```
 python train.py --gpus 1 --max_epochs 10 --model simple --batch-size 128 --task cifar --lr 1e-3 --momentum 0.9 --warmup 0 --cooldown 0 --precision 16 --terminate_on_nan --num-classes 10 --loss ce --embedding-dim 120
```

Removi `--normalize`; a perda foi bem menor, mas a acurácia não mostrou melhora. Foi pior. A acurácia de validação ficou estável, no entanto.

https://github.com/vitalwarley/research/blob/24debc65818b80150fde7665654804e1fde4c18e/model.py#L191-L192

![2022-04-03-162933_364x742_scrot](https://user-images.githubusercontent.com/6365065/161444777-53bde517-01fb-496a-93d7-0fbbb4699d7d.png)

---

Depois [atualizei](https://github.com/vitalwarley/research/commit/) o código pra se adequar ao original, mas tudo piorou.

**Comment by vitalwarley** [comment]
`--weight-decay` estava diferente (devia ser 0 em vez de 1e-4), mas também não foi a causa.

**Comment by vitalwarley** [comment]
Descobri o erro. Eu calculei a perda usando as `embeddings` (saída da penúltima camada: linear + relu) no lugar dos logits (saídos da última camada: linear apenas).

```
 Accuracy of my own SimpleModel trained with PTL on the 10000 test images: 62 %   
```

Treinei por 10 épocas.

**Comment by vitalwarley** [comment]
Reabri porque ainda vou treinar com meu modelo.

https://github.com/vitalwarley/research/blob/6b1f371e9c507a10a51ec3cb8957c6866e1dadbf/model.py#L210-L213

Será que para a perda ArcFace também devo usar os logits?

https://github.com/vitalwarley/research/blob/6b1f371e9c507a10a51ec3cb8957c6866e1dadbf/model.py#L190-L203

- `embeddings` vem do backbone, que vem sendo uma resnet pré-treinada.
- `logits` é a saída da última camada

https://github.com/vitalwarley/research/blob/6b1f371e9c507a10a51ec3cb8957c6866e1dadbf/model.py#L69-L86

Na implementação original a saída do [backbone](https://github.com/deepinsight/insightface/blob/eca1d9a6cd25653067e7293e27b8a2d0d2cd5415/recognition/arcface_torch/backbones/iresnet.py) lá é [512](https://github.com/deepinsight/insightface/blob/eca1d9a6cd25653067e7293e27b8a2d0d2cd5415/recognition/arcface_torch/configs/ms1mv3_r100.py#L12).

```python
# https://github.com/deepinsight/insightface/blob/eca1d9a6cd25653067e7293e27b8a2d0d2cd5415/recognition/arcface_torch/train.py#L79-L81
module_partial_fc = PartialFC(
    margin_loss, cfg.embedding_size, cfg.num_classes, 
    cfg.sample_rate, cfg.fp16)
```

**PartialFC** é basicamente uma camada densamente conectada, mas que atualiza as classes de forma esparsa (via amostragem parcial das classes) em vez de integralmente. É importante para problemas onde o número de classes é muito alto. Ver detalhes no [paper](https://arxiv.org/abs/2203.15565).

A instanciação do backbone

 ```python
# https://github.com/deepinsight/insightface/blob/eca1d9a6cd25653067e7293e27b8a2d0d2cd5415/recognition/arcface_torch/train.py#L58-L60
 backbone = get_model(
      cfg.network, dropout=0.0, fp16=cfg.fp16, num_features=cfg.embedding_size
  ).cuda()
```

que é usado para obter as `embeddings` (n_samples, 512).

```python
https://github.com/deepinsight/insightface/blob/eca1d9a6cd25653067e7293e27b8a2d0d2cd5415/recognition/arcface_torch/train.py#L133-L134
local_embeddings = backbone(img)
loss: torch.Tensor = module_partial_fc(local_embeddings, local_labels, opt)
```

Essas `embeddings`, por sua vez, é multiplicada com [pesos](https://github.com/deepinsight/insightface/blob/eca1d9a6cd25653067e7293e27b8a2d0d2cd5415/recognition/arcface_torch/partial_fc.py#L76-L88) de shape (n_sampled_classes, 512).

```python
https://github.com/deepinsight/insightface/blob/eca1d9a6cd25653067e7293e27b8a2d0d2cd5415/recognition/arcface_torch/partial_fc.py#L204-L214
with torch.cuda.amp.autocast(self.fp16):
    norm_embeddings = normalize(embeddings)
    norm_weight_activated = normalize(self.weight_activated)
    # (*, 512) * (n_classes?, 512) = (*, n_classes)
    logits = linear(norm_embeddings, norm_weight_activated)
if self.fp16:
    logits = logits.float()
logits = logits.clamp(-1, 1)

logits = self.margin_softmax(logits, labels)
loss = self.dist_cross_entropy(logits, labels)
return loss
```

No [paper](https://arxiv.org/abs/1801.07698) do ArcFace temos `logits` de dimensão final `n`, que identifica o número de classes. No entanto, a [página](https://kevinmusgrave.github.io/pytorch-metric-learning/losses/#arcfaceloss) da função que usei diz para usar as embeddings... 

> embedding_size: The size of the embeddings that you pass into the loss function. For example, if your batch size is 128 and your network outputs 512 dimensional embeddings, then set embedding_size to 512.

Acontece que dentro da [classe](https://github.com/KevinMusgrave/pytorch-metric-learning/blob/39a03468e15837bea31bedffcefde136f4a86b4b/src/pytorch_metric_learning/losses/large_margin_softmax_loss.py#L14) tudo é feito adequadmente. Apesar da confusão, meu código aparenta estar correto.

**Comment by vitalwarley** [comment]
![2022-04-03-211422_392x1046_scrot](https://user-images.githubusercontent.com/6365065/161455290-6e51dfc2-7270-4cf2-b1f2-a29f61f1a565.png)

Meu modelo com backbone `resnet34` e perde `ce`.

**Comment by vitalwarley** [comment]
` --loss arcface`

![2022-04-03-214133_320x881_scrot](https://user-images.githubusercontent.com/6365065/161456657-4bffe7ec-01af-42f6-85a7-efd6c210e3b3.png)

https://github.com/vitalwarley/research/blob/33987a66e6286eaca809719610d36df8ebd14357/model.py#L213-L215

Adicionei a linha 215 :)
